src/ripple/rpc/impl/ShardArchiveHandler.cpp

```diff
@@ -0,0 +1,245 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012-2014 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/app/misc/NetworkOPs.h>
+#include <ripple/basics/Archive.h>
+#include <ripple/core/ConfigSections.h>
+#include <ripple/nodestore/DatabaseShard.h>
+#include <ripple/rpc/ShardArchiveHandler.h>
+
+#include <memory>
+
+namespace ripple {
+namespace RPC {
+
+using namespace boost::filesystem;
+using namespace std::chrono_literals;
+
+ShardArchiveHandler::ShardArchiveHandler(Application& app, bool validate)
+    : app_(app)
+    , validate_(validate)
+    , downloadDir_(get(app_.config().section(
+        ConfigSection::shardDatabase()), "path", "") + "/download")
+    , timer_(app_.getIOService())
+    , j_(app.journal("ShardArchiveHandler"))
+{
+    assert(app_.getShardStore());
+}
+
+ShardArchiveHandler::~ShardArchiveHandler()
+{
+    timer_.cancel();
+    for (auto const& ar : archives_)
+        app_.getShardStore()->removePreShard(ar.first);
+
+    if (!is_directory(downloadDir_))
+        return;
+    try
+    {
+        remove_all(downloadDir_);
+    }
+    catch (std::exception const& e)
+    {
+        JLOG(j_.error()) <<
+            "Remove " << downloadDir_.string() <<
+            ": Exception, " << e.what();
+    }
+}
+
+bool
+ShardArchiveHandler::init()
+{
+    if (!app_.getShardStore())
+    {
+        JLOG(j_.error()) <<
+            "No shard store available";
+        return false;
+    }
+    if (downloader_)
+    {
+        JLOG(j_.error()) <<
+            "Already initialized";
+        return false;
+    }
+    downloader_ = std::make_shared<SSLHTTPDownloader>(
+        app_.getIOService(), j_);
+    return downloader_->init(app_.config());
+}
+
+bool
+ShardArchiveHandler::add(std::uint32_t shardIndex, parsedURL&& url)
+{
+    assert(downloader_);
+    auto it {archives_.find(shardIndex)};
+    if (it != archives_.end())
+    {
+        return url.scheme == it->second.scheme &&
```

Comment:
might make sense to add `operator ==` for `parsedURL`, but not a big deal since this is the only comparison in our code at this point.